ECIPs/ECIP-1027.md

```diff
@@ -0,0 +1,218 @@
+part of the network is used a test playground for new features of EVM. A business can apply custom optimization and 
+additional features (OPCODES and Predefined Contracts) for their needs on their Sidechain. It could move most of 
+politics aside, as sides can choose own subset of Sidechains for their needs.
+
+Besides EVM it allows to use different type of VM for particular Sidechain, such as eWASM, WASM, JVM, etc
+
+More to that, it’s possible to implement Sidechain with a totally new execution structure, which can be optimized just 
+for one particular task, such as Token Transfer. Token Transfer Sidechain doesn’t need most of EVM features, therefore 
+it can be implemented in ways similar to Bitcoin UTXO, Lighting/Raiden network, or can be used together with 
+Zk-SNARKs, etc.
+
+A particular Sidechain is free to use own subset of Sidechain, and practically allows to use Three of Chains, based 
+on task or conditions. That could be a long living chain or ephemeral chain, global chain or geo-bounded chain 
+suitable for local network of IoT devices.
+
+# Alternatives to proposed approach
+
+ - _Address Space Sharding_ [6] - where a peer keeps only part of address space required for processing. This solution can 
+   be used together with proposed scheme and it gives additional joint for scalability.
+ - _Raiden Network_ [7] - basically it’s a Sidechain in terms of this proposal, and can be considered as one of the possible 
+   implementations of a Sidechain
+   
+# Implementation
+
+## Checkpoint Contract Interface
+
+Minimal interface usable for a private managed chain
+
+```
+makeCheckpoint(uint64 height, uint256 difficulty, uint256 blockHash, uint256 parentHash, uint256 stateHash)
+
+getLastCheckpoint() 
+   returns (uint64 height, uint256 difficulty, uint256 blockHash, uint256 parentHash, uint256 stateHash)
+```
+
+It’s possible that for some cases it will be required to keep other data within the checkpoint, at this case it’s 
+possible to extend parameters list with this new details. But for better compatibility and for standardization, it’s 
+suggested to introduce custom methods to manage other details associated with the checkpoint
+
+## Extended interface
+
+```
+getHeight() 
+    returns (uint64 height)
+  
+getBlocksCount(uint64 height) 
+    returns (uint8 count)
+    
+getCheckpoint(uint64 height, uint8 index) 
+    returns (uint256 difficulty, uint256 blockHash, uint256 parentHash, uint256 stateHash)
+```
+
+## Geth Enhancements
+
+Geth should have support for multiple consensus mechanisms, and a configuration for multiple chains. Part of this 
+is already implemented, and we can have multiple custom/private chains with Geth starting at version 3.4.0. Similar
+features are already implemented in Parity.
+
+It will also require improvements in user interface, new commands, configuration options and documentation. None of 
+this is a breaking change
```

Comment:
It sounds like C1 and C2 are done as federated sidechains (like RSK on bitcoin - is that correct) but I'm not sure I understand C3. Does something comparable exist already?